File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/abstract_factory/lab/ConventionalLettuce.m
% ConventionalLettuce.m

classdef ConventionalLettuce < Lettuce
	properties
		NumberOfPesticides % int
	end

	methods
		% constructor:
		function theConventionalLettuce = ConventionalLettuce(aConventionalLettuceOrAPrice, aNumberOfPesticides)
			if nargin == 2
				% ConventionalLettuce(aPrice, aNumberOfPesticides):
				localAConventionalLettuceOrAPrice = aConventionalLettuceOrAPrice;
				localANumberOfPesticides          = aNumberOfPesticides;
			elseif nargin == 1
				if isa(aConventionalLettuceOrAPrice, 'ConventionalLettuce')
					% ConventionalLettuce(aConventionalLettuce):
					localAConventionalLettuceOrAPrice = aConventionalLettuceOrAPrice;
					localANumberOfPesticides          = aConventionalLettuceOrAPrice.NumberOfPesticides;
				else
					% ConventionalLettuce(aPrice):
					localAConventionalLettuceOrAPrice = aConventionalLettuceOrAPrice;
					localANumberOfPesticides          = 0; % the default value.
				end
			else
				error('Invalid number of input parameters.');
			end

			theConventionalLettuce@Lettuce(localAConventionalLettuceOrAPrice);

			theConventionalLettuce.NumberOfPesticides = localANumberOfPesticides;
		end

		function result = clone(theConventionalLettuce)
			result = ConventionalLettuce(theConventionalLettuce);
		end
	end
end

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/clone/lab/conventional_lettuce.cpp
// conventional_lettuce.cpp

#include "conventional_lettuce.h"

conventional_lettuce::conventional_lettuce(double a_price, int a_number_of_pesticides)
: lettuce(a_price)
{
	my_number_of_pesticides = a_number_of_pesticides;
}

conventional_lettuce::conventional_lettuce(const conventional_lettuce & other)
: lettuce(other)
{
	my_number_of_pesticides = other.my_number_of_pesticides;
}

conventional_lettuce::~conventional_lettuce()
{
}

conventional_lettuce* conventional_lettuce::clone() const
{
    return new conventional_lettuce(*this);
}
